torch/cuda/_memory_viz.py

```diff
@@ -309,36 +310,59 @@ def find_segment(addr):
             format(d)
     return out.getvalue()
 
-def trace_plot(data, device=None, plot_segments=False):
-
-    string_table = []
-    suffix_table = []
+class PlotWriter:
+    def __init__(self):
+        string_table: List[str] = []
+        suffix_table: List[Tuple[int, int]] = []
+
+        elements = []
+        actions: List[int] = []
+
+        initially_allocated: List[int] = []
+
+        @cache
+        def intern_str(s):
+            string_table.append(s)
+            return len(string_table) - 1
+
+        @cache
+        def intern_suffix(sid, restid):
+            suffix_table.append((sid, restid))
+            return len(suffix_table) - 1
+
+
+        def intern_stack(frames):
+            sids = [intern_str(f) for f in frames]
+            next_id = None
+            for sid in reversed(sids):
+                next_id = intern_suffix(sid, next_id)
+            return next_id
+
+        def add_element(size, lines):
+            elements.append({'size': size, 'info': intern_stack(lines)})
+            return len(elements) - 1
+
+        def to_html():
+            r = {
+                'actions': actions,
+                'elements': elements,
+                'suffix_table': suffix_table,
+                'string_table': string_table,
+                'initially_allocated': list(reversed(initially_allocated)),
+            }
+            plot_data = json.dumps(r)
+            return _memory_over_time_template.replace('$PLOT_DATA', plot_data)
 
-    elements = []
-    actions = []
+        self.add_element = add_element
```

Comment:
Is this because of `cache`? Or guarding `elements`? Or something else?

Reply:
Because of `cache`: it gets awkward to use the decorator on methods (I think by default it accidentally captures the object infinitely). It's non-idiomatic, but really only used here so we can reuse the suite of methods across the two formats.

For unit tests, after we get the initial functionality in, we can at least test that the export path produces a validly formatted html file.